File: step-by-step-directions-from-a-binary-tree-node-to-another.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
string start_to_LCA = "";
string LCA_to_end = "";
bool start_to_Node(TreeNode* lca,int start)
{
    if(!lca) return 0;
    if(lca->val==start) return 1;

    if(lca->left && start_to_Node(lca->left,start)==1) 
    {
        start_to_LCA += "U";
        return 1;
    }

    if(lca->right && start_to_Node(lca->right,start)==1)
    {
        start_to_LCA += "U";
        return 1;
    }

    return 0;
}

bool LCA_to_End(TreeNode* lca,int end)
{
    if(!lca) return 0;
    if(lca->val==end) return 1;


    if(LCA_to_End(lca->left,end))
    {
        LCA_to_end += "L";
        return 1;
    }

     if(LCA_to_End(lca->right,end))
    {
        LCA_to_end += "R";
        return 1;
    }

    return 0;
    
}
TreeNode* solve(TreeNode* root,int p,int q)
{
    if(!root) return NULL;
    if(root->val==p || root->val==q) return root;

    TreeNode* left = solve(root->left,p,q);
    TreeNode* right = solve(root->right,p,q);

    if(left && right) return root;
    if(!left && right) return right;
    if(left && !right) return left;
    return NULL;
}


    string getDirections(TreeNode* root, int startValue, int destValue) {
    string ans = "";
    TreeNode* LCA = solve(root,startValue,destValue);

    // now find a path from start to LCA  -> "UUUU...."
    // then find the path from LCA to end Node

    start_to_Node(LCA,startValue);

    LCA_to_End(LCA,destValue);
    
    reverse(LCA_to_end.begin(),LCA_to_end.end());
    return start_to_LCA + LCA_to_end  ;




    


      


    }
};

File: step-by-step-directions-from-a-binary-tree-node-to-another.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
        
    def solve(self, root, start, path, res):
        if not root:
            return
        
        if root.val == start:
            res.append(path[:])
            return
        
        path.append('L')
        self.solve(root.left, start, path, res)
        path.pop()
        
        path.append('R')
        self.solve(root.right, start, path, res)
        path.pop()
        
    def getDirections(self, root: TreeNode, startValue: int, destValue: int) -> str:
        path1, path2 = [], []
        self.solve(root, startValue, [], path1)
        self.solve(root, destValue, [], path2)
        
        s1, s2 = path1[0], path2[0]
        
        i = 0
        while i < len(s1) and i < len(s2) and s1[i] == s2[i]:
            i += 1
        
        t1 = 'U' * (len(s1) - i)
        t2 = ''.join(s2[i:])
        
        return t1 + t2
